src/GroupwiseConv.cc

```diff
@@ -0,0 +1,841 @@
+        if (outProcess.getBias() == nullptr) {
+          requantizeOutputProcessingGConvAvx2<
+              false,
+              false,
+              Q_GRAN,
+              false,
+              FUSE_RELU,
+              16>(out, inp, block, ld_out, ld_in, r);
+        } else {
+          requantizeOutputProcessingGConvAvx2<
+              false,
+              false,
+              Q_GRAN,
+              true,
+              FUSE_RELU,
+              16>(out, inp, block, ld_out, ld_in, r);
+        }
+      }
+    }
+  }
+}
+
+} // namespace
+
+template <
+    typename packed_W,
+    typename outType,
+    typename processOutputType,
+    int SPATIAL_DIM>
+void fbgemmGroupwiseConv(
+    const conv_param_t<SPATIAL_DIM>& conv_param,
+    const std::uint8_t* activations,
+    std::int32_t a_zero_point,
+    std::int32_t* rowOffsetBuf,
+    packed_W& packed_weights,
+    outType* out,
+    int32_t* outBuffer,
+    const processOutputType& outProcess,
+    int thread_id,
+    int num_threads) {
+  return fbgemmGroupwiseConvBase_<
+      packed_W,
+      outType,
+      processOutputType,
+      SPATIAL_DIM>(
+      conv_param,
+      activations,
+      a_zero_point,
+      rowOffsetBuf,
+      packed_weights,
+      out,
+      outBuffer,
+      outProcess,
+      thread_id,
+      num_threads);
+}
+
+template <
+    typename packed_W,
+    typename outType,
+    bool FUSE_RELU,
+    QuantizationGranularity Q_GRAN,
+    int SPATIAL_DIM>
+void fbgemmGroupwiseConv(
+    const conv_param_t<SPATIAL_DIM>& conv_param,
+    const std::uint8_t* activations,
+    std::int32_t a_zero_point,
+    std::int32_t* rowOffsetBuf,
+    packed_W& packed_weights,
+    outType* out,
+    int32_t* outBuffer,
+    const ReQuantizeOutput<FUSE_RELU, Q_GRAN>& outProcess,
+    int thread_id,
+    int num_threads) {
+  typedef ReQuantizeOutput<FUSE_RELU, Q_GRAN> processOutputType;
+
+  if (!fbgemmOptimizedGConv<SPATIAL_DIM>(conv_param) ||
+      (!fbgemmHasAvx512Support() && !fbgemmHasAvx2Support())) {
+    return fbgemmGroupwiseConvBase_<
+        packed_W,
+        outType,
+        processOutputType,
+        SPATIAL_DIM>(
+        conv_param,
+        activations,
+        a_zero_point,
+        rowOffsetBuf,
+        packed_weights,
+        out,
+        outBuffer,
+        outProcess,
+        thread_id,
+        num_threads);
+  }
+
+  int MB = conv_param.MB;
+  int H_in = conv_param.IN_DIM[0];
+  int W_in = conv_param.IN_DIM[1];
+  int W_out = conv_param.OUT_DIM[1];
+  int G = conv_param.G;
+  int K_per_G = conv_param.OC / G;
+  int C_per_G = conv_param.IC / G;
+  int R = conv_param.K[0];
+  int S = conv_param.K[1];
+  int ih_iw = H_in * W_in;
+
+#ifdef FBGEMM_MEASURE_TIME_BREAKDOWN
+  chrono::time_point<chrono::high_resolution_clock> t_begin, t_end;
+#endif
+
+  static_assert(SPATIAL_DIM == 2, "3D conv not supported yet");
+
+  assert(G % 8 == 0);
+  // Overall loop order
+  // for each MB // i0
+  //  for each G
+  //   for each H // i1
+  //    for each g1
```

Comment:
what is g1 and g2?
what is g in gRSck?

Reply:
Sorry I was sloppy when I was writing this comment. Will update soon. Upper case is outer loop for that dimension and lower case is inner loops for that dimension. For group dimension, we have 3 levels of loop, G, g1, and g2. g2 is for vectorization (only used when we don't have enough channels per group to fill the whole SIMD register), and g1 is for vectorizing the computation of row_offset (we need to work on 8 groups at a time if we vectorize row_offset computation across groups).
R = vertical filter size (usually 1 or 3)
S = horizontal filter size (usually 1 or 3)